[MLIR][Presburger] Implement matrix inverse

Abhinav271828 opened the pull request:
Shift the `determinant()` function from LinearTransform to Matrix.
Implement a FracMatrix class, inheriting from Matrix<Fraction>, for inverses.
Implement inverse for FracMatrix and intInverse for IntMatrix.

Abhinav271828 committed:
Template Matrix to Matrix<T> (for MPInt and Fraction) with explicit instantiation
Duplicate makeMatrix to makeIntMatrix and makeFracMatrix

Implement arithmetic operations for Fraction for compatibility

Abhinav271828 committed:
Template Matrix to Matrix<T> (for MPInt and Fraction) with explicit instantiation
Duplicate makeMatrix to makeIntMatrix and makeFracMatrix

Implement arithmetic operations for Fraction for compatibility

Abhinav271828 committed:
Template Matrix to Matrix<T> (for MPInt and Fraction) with explicit instantiation
Duplicate makeMatrix to makeIntMatrix and makeFracMatrix

Implement arithmetic operations for Fraction for compatibility

Abhinav271828 committed:
Template Matrix to Matrix<T> (for MPInt and Fraction) with explicit instantiation
Duplicate makeMatrix to makeIntMatrix and makeFracMatrix

Implement arithmetic operations for Fraction for compatibility

Abhinav271828 committed:
Template Matrix to Matrix<T> (for MPInt and Fraction) with explicit instantiation
Duplicate makeMatrix to makeIntMatrix and makeFracMatrix

Implement arithmetic operations for Fraction for compatibility

Abhinav271828 committed:
Template Matrix to Matrix<T> (for MPInt and Fraction) with explicit instantiation
Duplicate makeMatrix to makeIntMatrix and makeFracMatrix

Implement arithmetic operations for Fraction for compatibility

Abhinav271828 committed:
Template Matrix to Matrix<T> (for MPInt and Fraction) with explicit instantiation
Duplicate makeMatrix to makeIntMatrix and makeFracMatrix

Implement arithmetic operations for Fraction for compatibility

Abhinav271828 committed:
Template Matrix to Matrix<T> (for MPInt and Fraction) with explicit instantiation
Duplicate makeMatrix to makeIntMatrix and makeFracMatrix

Implement arithmetic operations for Fraction for compatibility

Abhinav271828 committed:
Fix rebase conflict

Abhinav271828 committed:
Add static assert

Abhinav271828 committed:
Add static assert

Abhinav271828 committed:
Add static assert

Abhinav271828 committed:
Add static assert

Abhinav271828 committed:
Fix comment

Abhinav271828 committed:
Shift changes to Fraction

Abhinav271828 committed:
Shift changes to Fraction

Abhinav271828 committed:
Update documentation and remove extraneous include

Abhinav271828 committed:
Update documentation and remove extraneous include

Abhinav271828 committed:
Make compatible with fraction patch

Abhinav271828 committed:
Revert "Make compatible with fraction patch"

This reverts commit 2871b0127a3cbc8d2a4cf294e731aade1ed56424.

Abhinav271828 committed:
Revert "Make compatible with fraction patch"

This reverts commit 2871b0127a3cbc8d2a4cf294e731aade1ed56424.

Abhinav271828 committed:
Remove duplicate definitions

Abhinav271828 committed:
Remove duplicate definitions

Abhinav271828 committed:
Update upstream branch

Abhinav271828 committed:
Fix duplicates and use increment operators for Fraction

Abhinav271828 committed:
Fix duplicates and use increment operators for Fraction

Abhinav271828 committed:
Fix duplicates and delete extra files

Abhinav271828 committed:
Fix duplicates and delete extra files

Abhinav271828 committed:
Fix comment

Abhinav271828 committed:
Create IntMatrix for hermite normal form and row normalisation

Abhinav271828 committed:
Create IntMatrix for hermite normal form and row normalisation

Abhinav271828 committed:
Put opening braces on same line

Abhinav271828 committed:
Merge branch 'llvm:main' into first_patch

Abhinav271828 committed:
Inherit Matrix<Fraction> to FracMatrix and define inverse
Fix reduce() method in Fraction

Abhinav271828 committed:
Inherit Matrix<Fraction> to FracMatrix and define inverse
Fix reduce() method in Fraction

Abhinav271828 committed:
Add test for matrix inverse

Abhinav271828 committed:
Add test for matrix inverse

Abhinav271828 committed:
Formatting

Abhinav271828 committed:
Formatting

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -390,4 +390,68 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+}
+
+FracMatrix FracMatrix::identity(unsigned dimension) {
+  FracMatrix matrix(dimension, dimension);
```
You can use Matrix<T>::identity in this implementation

Abhinav271828 committed:
Shift determinant from LinearTransform to Matrix<T>

Abhinav271828 committed:
Shift determinant from LinearTransform to Matrix<T>

Abhinav271828 committed:
Implement integer inverse for integer matrix

Abhinav271828 committed:
Implement integer inverse for integer matrix

Abhinav271828 committed:
Formatting

Abhinav271828 committed:
Formatting

Abhinav271828 committed:
Merge github.com:llvm/llvm-project into upstream

Superty commented:
By the way, what happens if the inverse doesn't exist?

Superty commented:
By the way, what happens if the inverse doesn't exist?

Abhinav271828 committed:
Revert "Inherit Matrix<Fraction> to FracMatrix and define inverse"

This reverts commit 80f1d9654f09667679f7565101fbb62ce2d6aaa6.

Abhinav271828 committed:
Revert "Inherit Matrix<Fraction> to FracMatrix and define inverse"

This reverts commit 80f1d9654f09667679f7565101fbb62ce2d6aaa6.

Abhinav271828 committed:
Fix reduce and add Fraction tests

Abhinav271828 committed:
Fix reduce and add Fraction tests

Abhinav271828 committed:
Formatting

Abhinav271828 committed:
Formatting

Abhinav271828 committed:
Fix CMake file

Abhinav271828 committed:
Fix CMake file

Abhinav271828 committed:
Formatting

Abhinav271828 committed:
Sync up

Abhinav271828 committed:
Change assert to expect

Abhinav271828 committed:
Revert "Revert "Inherit Matrix<Fraction> to FracMatrix and define inverse""

Abhinav271828 committed:
Revert "Revert "Inherit Matrix<Fraction> to FracMatrix and define inverse""

Abhinav271828 committed:
Fix reduce

Abhinav271828 committed:
Template Matrix to Matrix<T> (for MPInt and Fraction) with explicit instantiation
Duplicate makeMatrix to makeIntMatrix and makeFracMatrix

Implement arithmetic operations for Fraction for compatibility

Abhinav271828 committed:
Sync

Abhinav271828 committed:
Add newlines

Abhinav271828 committed:
Merge branch 'main' into first_patch

Abhinav271828 committed:
Use Matrix::identity() for FracMatrix::identity()

Abhinav271828 committed:
Merge branch 'first_patch' of github.com:Abhinav271828/mlir-barvinok into first_patch

Abhinav271828 committed:
Formatting

Abhinav271828 committed:
Make inverse functions return optionals

Abhinav271828 committed:
Revert "Formatting"

This reverts commit 664a7d964ebe59d85c48c475858ca2082133080c.

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -390,4 +463,83 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+  unsigned dim = getNumRows();
+
+  // Construct the augmented matrix [M | I]
+  FracMatrix augmented(dim, dim + dim);
+  for (unsigned i = 0; i < dim; i++) {
+    augmented.fillRow(i, 0);
+    for (unsigned j = 0; j < dim; j++)
+      augmented(i, j) = at(i, j);
+    augmented(i, dim + i).num = 1;
+    augmented(i, dim + i).den = 1;
+  }
```
I think it's cleaner to just maintain a copy of `this` and the inverse as separate matrices. You can just assign to 1 btw. and filling 0 should be by default, check the constructor doc.

Abhinav271828 replied:
What do you mean maintain a copy of `this`? I'm a little confused.

Superty replied:
The matrix to be inverted, the one corresponding to the object the function is called on

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -390,4 +463,83 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+    if (augmented(i, i) == Fraction(0, 1))
+      for (unsigned j = i + 1; j < dim; j++)
+        if (augmented(j, i) != Fraction(0, 1)) {
+          augmented.addToRow(i, augmented.getRow(j), Fraction(1, 1));
```
there should be an `addRows` or something to do this directly

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -390,4 +463,83 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+  }
+  Fraction a, b;
+  for (unsigned i = 0; i < dim; i++) {
+    if (augmented(i, i) == Fraction(0, 1))
```
prefer early exits instead of deep nesting. if the body is big and especially if it contains braces, then use braces for it

Abhinav271828 replied:
This condition is only for the `for` loop, it doesn't include the entire following body. can we still use early exit?

Superty replied:
guess not

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -390,4 +463,83 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+  for (unsigned i = 0; i < dim; i++) {
+    if (augmented(i, i) == Fraction(0, 1))
+      for (unsigned j = i + 1; j < dim; j++)
+        if (augmented(j, i) != Fraction(0, 1)) {
```
doesn't fraction have comparisons to int defined?

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -390,4 +463,83 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+      a = augmented(j, i);
+      // Rj -> Rj - (b/a)Ri
+      augmented.addToRow(j, augmented.getRow(i), -a / b);
+      // Now (Rj)i = 0
```
please write full sentences for documentation throughout and explain what each step is doing in this main loop.

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -283,12 +326,43 @@ template <typename T> bool Matrix<T>::hasConsistentState() const {
+      n++;
+    }
+
+    determinant = determinant + at(0, i) * sign * cofactor.determinant();
```
please don't use this exponential recursive implementation. it's going to be pretty slow and unbounded recursion is does not conform to LLVM standards.  do you need it for both integers and fractions? try to use the gaussian elimination for this.

Superty commented on mlir/include/mlir/Analysis/Presburger/Matrix.h:
```diff
@@ -241,6 +245,32 @@ class IntMatrix : public Matrix<MPInt>
+      : Matrix<Fraction>(rows, columns, reservedRows, reservedColumns){};
+
+  FracMatrix(Matrix<Fraction> m)
+      : Matrix<Fraction>(m.getNumRows(), m.getNumColumns(),
+                         m.getNumReservedRows(), m.getNumReservedColumns()) {
+    for (unsigned i = 0; i < m.getNumRows(); i++)
+      for (unsigned j = 0; j < m.getNumColumns(); j++)
+        at(i, j) = m(i, j);
```
`: Matrix<Fraction(m)`?

Abhinav271828 committed:
Merge branch 'main' into first_patch

Abhinav271828 committed:
Merge branch 'first_patch' of github.com:Abhinav271828/mlir-barvinok into first_patch

Abhinav271828 committed:
Refactor addToRow call

Abhinav271828 committed:
Comparison with integers

Abhinav271828 committed:
Fix assignment and initialisation in inverse

Abhinav271828 committed:
Inverse documentation

Abhinav271828 committed:
Simplify constructors

Abhinav271828 committed:
Make determinant recursive

Abhinav271828 committed:
Assert square matrix in det

Abhinav271828 committed:
Maintain copy of inverse

Abhinav271828 committed:
Formatting

Superty commented on mlir/include/mlir/Analysis/Presburger/Matrix.h:
```diff
@@ -36,6 +36,8 @@ namespace presburger {
 /// This class only works for the types MPInt and Fraction, since the method
 /// implementations are in the Matrix.cpp file. Only these two types have
 /// been explicitly instantiated there.
+/// Note that the determinant() functions (for both integers and fractions)
+/// assume that the matrix is square.
```
Add these as documentation for the determinant function please, not the class.

Superty commented on mlir/include/mlir/Analysis/Presburger/Matrix.h:
```diff
@@ -239,6 +236,36 @@ class IntMatrix : public Matrix<MPInt> {
   /// Returns the GCD of the columns of the specified row.
   MPInt normalizeRow(unsigned row);
+
+  // Return the integer inverse of the matrix, leaving the calling object
+  // unmodified.
```
I don't know what integer inverse is; please document what is being returned exactly and when the optional is empty.

Superty commented on mlir/include/mlir/Analysis/Presburger/Matrix.h:
```diff
@@ -239,6 +236,36 @@ class IntMatrix : public Matrix<MPInt> {
+  // unmodified.
+  std::optional<IntMatrix> integerInverse();
+
+  // Compute the determinant of the matrix by converting it to row echelon
+  // form and then taking the product of the diagonal.
```
the algorithm doesn't matter here. the header level documentation is for the user to decide whether to use the function. Just mention a complexity estimate here.

Superty commented on mlir/include/mlir/Analysis/Presburger/Matrix.h:
```diff
@@ -239,6 +236,36 @@ class IntMatrix : public Matrix<MPInt> {
+  static FracMatrix identity(unsigned dimension);
+
+  // Return the inverse of the matrix, leaving the calling object unmodified.
+  std::optional<FracMatrix> inverse();
```
if the calling object is unmodified then add the const keyword

`std::optional<FracMatrix> inverse() const;`

Superty commented on mlir/include/mlir/Analysis/Presburger/Matrix.h:
```diff
@@ -239,6 +236,36 @@ class IntMatrix : public Matrix<MPInt> {
+
+  // Compute the determinant of the matrix by converting it to row echelon
+  // form and then taking the product of the diagonal.
+  Fraction determinant();
```
add const keyword.

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -432,4 +432,138 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+MPInt IntMatrix::determinant() {
+  unsigned r = getNumRows();
+  unsigned c = getNumColumns();
+  assert(r == c);
```
add an assert message please

`assert(r == c && "determinant can only be calculated for square matrices!");`

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -432,4 +432,138 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+}
+
+MPInt IntMatrix::determinant() {
+  unsigned r = getNumRows();
+  unsigned c = getNumColumns();
```
please use full var names like `nRows`, `nCols`

Superty replied:
if they are already member variables in Matrix, you can use them directly

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -432,4 +432,138 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+  FracMatrix m(r, c);
+  for (unsigned i = 0; i < r; i++)
+    for (unsigned j = 0; j < c; j++)
+      m.at(i, j) = Fraction(at(i, j), 1);
```
isn't there an assignment operator from MPInt to Fraction?

Abhinav271828 replied:
it's ambiguous because both
`Fraction(const MPInt &oNum, const MPInt &oDen = MPInt(1))`
and
`Fraction(const MPInt &num, int64_t den = 1) : Fraction(num, MPInt(den))`
apply. should I make one of these not have a default second argument?

Superty replied:
Fixed upstream

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -432,4 +432,138 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+  unsigned c = getNumColumns();
+  assert(r == c);
+
+  FracMatrix m(r, c);
+  for (unsigned i = 0; i < r; i++)
+    for (unsigned j = 0; j < c; j++)
+      m.at(i, j) = Fraction(at(i, j), 1);
```
directly copy construct `m = *this`?

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -432,4 +432,138 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+
+std::optional<IntMatrix> IntMatrix::integerInverse() {
+  Fraction det = Fraction(determinant(), 1);
+  FracMatrix newMat(getNumRows(), getNumColumns());
+  for (unsigned i = 0; i < getNumRows(); i++)
+    for (unsigned j = 0; j < getNumColumns(); j++)
+      newMat(i, j) = Fraction(at(i, j), 1);
```
copy construct directly

Abhinav271828 replied:
it doesn't work, we don't have a cast from IntMatrix to FracMatrix. shall I write one?

Superty replied:
sure. make it an explicit constructor

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -432,4 +432,138 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+  return det.getAsInteger();
+}
+
+std::optional<IntMatrix> IntMatrix::integerInverse() {
```
add some doc comments to explain what's going on

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -432,4 +432,138 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+  // as two separate matrices, M and I.
+  FracMatrix modified(dim, dim);
+  FracMatrix inverse = FracMatrix::identity(dim);
+  for (unsigned i = 0; i < dim; i++) {
+    for (unsigned j = 0; j < dim; j++)
+      modified(i, j) = at(i, j);
+  }
+
```
copy construct

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -432,4 +432,138 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+  return Matrix::identity(dimension);
+}
+
+Fraction FracMatrix::determinant() {
```
is it just me or has gaussian elimination been implemented twice here, once for det and once for inverse?

Superty replied:
I think the best thing for now is to just make one function that returns both (the inverse can be a pointer that will be updated if it is passed, i.e., is non-null)

superty reviewed (changes requested):
Added some more comments. Please bear with me here, I know that the review process is pretty long. It's quite common that this happens with the first substantial patch that someone sends. After this you will get a much better idea of how things are supposed to look and it should be faster from there.

Abhinav271828 committed:
Use copy constructor

Abhinav271828 committed:
Documentation for integerInverse

Abhinav271828 committed:
Miscellaneous

Abhinav271828 committed:
Refactor determinant and inverse into single function

Abhinav271828 committed:
Add cast from IntMatrix to FracMatrix

Superty commented on mlir/include/mlir/Analysis/Presburger/Matrix.h:
```diff
@@ -36,6 +36,7 @@ namespace presburger {
 /// This class only works for the types MPInt and Fraction, since the method
 /// implementations are in the Matrix.cpp file. Only these two types have
 /// been explicitly instantiated there.
+/// assume that the matrix is square.
```
leftover line.

Superty commented on mlir/include/mlir/Analysis/Presburger/Matrix.h:
```diff
@@ -239,6 +235,39 @@ class IntMatrix : public Matrix<MPInt> {
+  // For a matrix M, the integer inverse is the matrix M' such that
+  // M • M' = M' • M = det(M) • I.
+  // Assumes that the matrix is square.
+  MPInt determinant(IntMatrix* inverse = NULL) const;
```
NULL -> nullptr

Superty replied:
Just use `x` for multiplication instead of this symbol

Superty commented on mlir/include/mlir/Analysis/Presburger/Matrix.h:
```diff
@@ -239,6 +235,39 @@ class IntMatrix : public Matrix<MPInt> {
+  // Compute the determinant of the matrix (cubic time).
+  // Stores the inverse of the matrix in the pointer
+  // passed. The pointer is unchanged if the inverse
+  // does not exist.
```
might be making it explicit that inverse does not exist iff determinant is zero

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -432,4 +432,111 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+  FracMatrix m(*this);
+  
+  FracMatrix fracInverseM(nRows, nColumns);
+  FracMatrix* fracInverse = &fracInverseM;
+  Fraction detF = m.determinant(fracInverse);
```
just directly pass `&fracInverseM`. rename fracInverseM to fracInverse

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -432,4 +432,111 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+  // Now only diagonal elements are nonzero, but they are
+  // not necessarily 1. We normalise them.
+  for (unsigned i = 0; i < nRows; i++) {
+    a = m(i, i);
```
this is not necessary. you can write m(i, i) directly below and remove the outer braces on these loops

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -432,4 +432,111 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+  if (inverse) tempInv = FracMatrix::identity(nRows);
+
+  Fraction a, b;
+  // For each row in the matrix,
```
please give an overview at the top of the loop. We can see that the loop is over the rows of the matrix. instead of that, add high-level information like 

// Make the matrix into lower triangular form using gaussian elimination with row operations.
// If inverse is required, we apply more operations to turn the matrix into diagonal form and apply
// the same operations to the inverse matrix that was initially identity. Either way, the product of the
// diagonal elements is then the determinant.

Some duplication with the comments inside the loop is ok

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -432,4 +432,111 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+      tempInv.at(i, j) = tempInv.at(i, j) / a;
+  }
+
+  *inverse = tempInv;
```
If inverse. How was this not caught in the tests?

Superty replied:
I meant if the pointer is non-null

Abhinav271828 commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -432,4 +432,111 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+    // a nonzero row to it.
+      for (unsigned j = i + 1; j < nRows; j++)
+        if (m(j, i) != 0) {
+          m.addToRow(j, i, 1);
```
This isn't swapping, though; we're doing Rj = Rj + Ri.
Can this be expressed in terms of swaps somehow?

Superty replied:
swapping is also a valid row operation.

Abhinav271828 committed:
Change Matrix attributes to protected
Formatting

Abhinav271828 commented on mlir/include/mlir/Analysis/Presburger/Matrix.h:
```diff
@@ -239,6 +235,39 @@ class IntMatrix : public Matrix<MPInt> {
+  static FracMatrix identity(unsigned dimension);
+
+  // Compute the determinant of the matrix (cubic time).
+  // Stores the inverse of the matrix in the pointer
+  // passed. The pointer is unchanged if the inverse
+  // does not exist.
+  // Assumes that the matrix is square.
```
clang-format didn't affect the comments.

Abhinav271828 committed:
Formatting

Abhinav271828 committed:
Fix singularity condition in inverse

Abhinav271828 committed:
Merge github.com:llvm/llvm-project into upstream

Abhinav271828 committed:
Merge branch 'upstream' into first_patch

Abhinav271828 committed:
Fix MPInt to Fraction conversion

Abhinav271828 committed:
Fix nonzeroification

Abhinav271828 committed:
Change add to swap

Superty commented on mlir/unittests/Analysis/Presburger/MatrixTest.cpp:
```diff
@@ -248,3 +248,61 @@ TEST(MatrixTest, computeHermiteNormalForm) {
+      2, 2,
+      {{Fraction(0, 1), Fraction(1, 7)}, {Fraction(1, 1), Fraction(-2, 7)}});
+
+  FracMatrix invM(2, 2);
+  FracMatrix* inv = &invM;
+  mat.determinant(inv);
```
you can just pass &invM

Superty commented on mlir/unittests/Analysis/Presburger/MatrixTest.cpp:
```diff
@@ -248,3 +248,61 @@ TEST(MatrixTest, computeHermiteNormalForm) {
+TEST(MatrixTest, inverse) {
+  FracMatrix mat = makeFracMatrix(
+      2, 2,
+      {{Fraction(2, 1), Fraction(1, 1)}, {Fraction(7, 1), Fraction(0, 1)}});
```
can you format all matrices by rows and columns so that we can read them easily

Superty commented on mlir/unittests/Analysis/Presburger/MatrixTest.cpp:
```diff
@@ -248,3 +248,61 @@ TEST(MatrixTest, computeHermiteNormalForm) {
+
+TEST(MatrixTest, intInverse) {
+  IntMatrix mat = makeIntMatrix(2, 2, {{2, 1}, {7, 0}});
+  IntMatrix inverse = makeIntMatrix(2, 2, {{0, -1}, {-7, 2}});
```
why is this not a scale of the inverse above?

Abhinav271828 replied:
it is, it's scaled by -7, right?

Superty replied:
ah but shouldn't it be scaled by a positive number, the determinant?

Superty commented on mlir/unittests/Analysis/Presburger/MatrixTest.cpp:
```diff
@@ -248,3 +248,61 @@ TEST(MatrixTest, computeHermiteNormalForm) {
+  FracMatrix inverse = makeFracMatrix(
+      2, 2,
+      {{Fraction(0, 1), Fraction(1, 7)}, {Fraction(1, 1), Fraction(-2, 7)}});
+
```
add some tests that are 0x0, 0x1, 1x0, 1x1

Abhinav271828 replied:
0x1 and 1x0 will throw errors right?

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -432,4 +432,113 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+    }
+  }
+
+  // Now only diagonal elements are nonzero, but they are
+  // not necessarily 1. We normalise them.
+  for (unsigned i = 0; i < nRows; i++)
+    for (unsigned j = 0; j < nRows; j++)
+      tempInv.at(i, j) = tempInv.at(i, j) / m(i, i);
+
+  *inverse = tempInv;
```
this doesn't seem to be resolved. this should only be called if inverse is non-null

Superty replied:
// Now only diagonal elements of `m` are nonzero, but they are
  // not necessarily 1. To get the true inverse, we should normalize them and apply the same scale to
// the inverse matrix. For efficiency we skip scaling `m` and just scale `tempInv` appropriately.

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -432,4 +432,113 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+  if (detM == 0)
+    return MPInt(0);
+
+  *inverse = IntMatrix(nRows, nColumns);
```
if (inverse)

Superty commented on mlir/unittests/Analysis/Presburger/MatrixTest.cpp:
```diff
@@ -248,3 +248,61 @@ TEST(MatrixTest, computeHermiteNormalForm) {
+  IntMatrix inverse = makeIntMatrix(2, 2, {{0, -1}, {-7, 2}});
+
+  IntMatrix invM(2, 2);
+  IntMatrix* inv = &invM;
```
add some tests where inverse is nullptr

Abhinav271828 committed:
Add test for swapRow

Abhinav271828 committed:
Improve testing

Superty commented on mlir/unittests/Analysis/Presburger/MatrixTest.cpp:
```diff
@@ -248,3 +248,85 @@ TEST(MatrixTest, computeHermiteNormalForm) {
+  IntMatrix inv(2, 2);
+  mat.determinant(&inv);
+
+  for (unsigned i = 0; i < 2u; i++)
+    for (unsigned j = 0; j < 2u; j++)
+      EXPECT_EQ(inv(i, j), inverse(i, j));
+
```
I would suggest just adding a Matrix operator== and then using EXPECT_EQ on the matrices here. Either that or at least make a function to EXPECT_EQ two matrices (and their dimensions)

Abhinav271828 committed:
Function to test matrix equality

Abhinav271828 committed:
Formatting & tidying

Abhinav271828 committed:
Formatting

Superty commented on mlir/lib/Analysis/Presburger/Matrix.cpp:
```diff
@@ -432,4 +432,119 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
+      // First ensure that the diagonal
+      // element is nonzero, by swapping
+      // it with a nonzero row.
+      for (unsigned j = i + 1; j < nRows; j++)
```
Please add braces whenever the body contains another brace-enclosed block

Abhinav271828 committed:
Formatting
